rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,146 @@
+
+Currently it is difficult to accurately gauge the volume of data that is moving
+through Vector. It is difficult to query where data being sent out has come
+from.
+
+## Proposal
+
+### User Experience
+
+Global config options will be provided allowing the `service` tag and the
+`source` tag to be specified. For example:
+
+```yaml
+metrics:
+  tags:
+    service: service
+    source: input
+```
+
+The default will be to not emit these tags.
+
+### Implementation
+
+#### Metric tags
+
+*service* - to attach the service, we need to add a new meaning to Vector - *service*. Any sources that
+            receive data that could potentially be considered a service will need to indicate which field
+            means `service`.
+            This work has largely already been done with the LogNamespacing work, so it will be trivial to add
+            this new field.
+
+*source* - A new field will be added to the [Event metadata][event_metadata] - `Arc<ComponentId>` that will indicate the source
+           of the event.
+
+We will need to do an audit of all components to ensure the bytes emitted for the `component_received_event_bytes_total`
+and `component_sent_event_bytes_total` metrics are the estimated JSON size of the event.
+
+#### `component_received_event_bytes_total`
+
+This metric is emitted by the framework [here][source_sender], so it looks like the only change needed is
+to add the service tag.
+
+#### `component_sent_event_bytes_total`
+
+For stream based sinks this will typically be the byte value returned by `DriverResponse::events_sent`.
+
+Despite being in the [Component Spec][component_spec], not all sinks currently conform to this.
+
+As an example, from a cursory glance over a couple of sinks:
+
+The Amqp sink currently emits this value as the length of the binary data that is sent. By the time the data has
+reached the code where the `component_sent_event_bytes_total` event is emitted, that event has been encoded
+and the actual estimated JSON size has been lost. The sink will need to be updated so that when the event is
+encoded, the encoded event together with the pre-encoded JSON bytesize will be sent to the service where the
+event is emitted.
+
+The Kafka sink also currently sends the binary size, but it looks like the estimated JSON bytesize is easily
+accessible at the point of emitting, so would not need too much of a change.
+
+To ensure that the correct metric is sent in a type-safe manner, we will wrap the estimated JSON size in a
+newtype:
+
+```rust
+pub struct JsonSize(usize);
+```
+
+The `EventsSent` metric will only accept this type.
+
+## Rationale
+
+The ability to visualize data flowing through Vector will allow users to ascertain
+the effectiveness of the current use of Vector. This will enable users to
+optimise their configurations to make the best use of Vector's features.
+
+## Drawbacks
+
+The additional tags being added to the metrics will increase the cardinality of
+those metrics if they are enabled.
```

Comment:
Also curious if this breaks some of the event registration work we did with the need to add this new dynamic tag. cc/ @bruceg (who I also tagged as reviewer)

Reply:
It looks like it will.

Reply:
It absolutely will, ouch. Given that registered events have become a correctness issue due to expiring metrics, I think we will have to resolve that breakage. I see only one way out, and that is to introduce a registered event caching layer that is keyed on the source/service identifiers. Obviously, that adds more complexity and interacts negatively with expiring metrics.